analyses/tmb-compare-tcga/TMB_d3b_code/code/01_calculate_tmb_targetflexible_withbothexperstrt_and_cohort.py

```diff
@@ -0,0 +1,235 @@
+            lambda x: x["Start_Position"] - 1, axis=1
+        )
+        target_bed = target_dict.get(sample_target_identifier)
+        maf_within_target = pybedtools.BedTool.from_dataframe(grouped_df).intersect(
+            target_bed, u=True
+        )
+        mafdf_within_target = pybedtools.BedTool.to_dataframe(
+            maf_within_target, names=cols
+        )
+        count = mafdf_within_target.shape[0]
+        bed_length = calculate_bed_length(open(target_bed, "r"))
+        tmb = (count * 1000000) / bed_length
+        out_line = (
+            samplename
+            + "\t"
+            + exp_strategy
+            +"\t"
+            + cohort
+            + "\t"
+            + disease
+            + "\t"
+            + str(count)
+            + "\t"
+            + str(bed_length)
+            + "\t"
+            + str(tmb)
+        )
+        out.write(out_line + "\n")
+
+
+# This function takes in BED file and calculates the total length of the BED
+def calculate_bed_length(in_bed):
+    total_length = 0
+    for line in in_bed:
+        if line.split()[0] != "chrom":
+            try:
+                total_length += int(line.split()[2]) - int(line.split()[1])
+            except IndexError:
+                print("Check BED file formatting\n")
+    return total_length
+
+
+############### REading in input files #################################
+
+parser = argparse.ArgumentParser()
+parser.add_argument("-i", "--maf", required=True, help="path to the MAF file")
+parser.add_argument(
+    "-m", "--metadatafile", required=True, help="path to the metadata/histology file"
+)
+parser.add_argument("-o", "--outfilename", required=True, help="Out file name")
+parser.add_argument(
+    "-c",
+    "--configfile",
+    required=True,
+    help="calculate_tmb.cfg.txt file with columns for disease, samplename, variant types etc.",
+)
+parser.add_argument(
+    "-w",
+    "--targetconfig",
+    required=True,
+    help="File with experimental strategy  and path to BED file",
+)
+args = parser.parse_args()
+
+
+################################################################
+# Defining MAF columns and metadata spcific fields from config #
+
+needed_cols = [
+    "Chromosome",
+    "Start_Position",
+    "End_Position",
+    "Variant_Classification",
+    "Tumor_Sample_Barcode",
+]
+
+with open(args.configfile) as configlines:
+    for line in configlines.readlines():
+        if line.startswith("Variants_to_use_for_TMB="):
+            var_class = line.split("=")[1].split(",")
+        if line.startswith("disease_column="):
+            disease_col = line.split("=")[1].strip("\n")
+        if line.startswith("samplename_column="):
+            samplename_col = line.split("=")[1].strip("\n")
+        if line.startswith("cohort_column="):
+            cohort_col = line.split("=")[1].strip("\n")
+        if line.startswith("typeoftargetcolumn="):
+            typeoftargetcol = line.split("=")[1].strip("\n")
+
```

Comment:
This may be out of scope for this code, but if you use a standard format like JSON for the config file, this section can be very much simplified to something like this:

```
# this goes at the top
import json

# here
with open(args.configfile) as f:
    config = json.load(f)
    var_class = config["Variants_to_use_for_TMB"] # this will already be a list!
    disease_col = config["disease_column"]
    ...
```